File: solutions/python/food-chain/1/food_chain.py
animals = [
    ("fly", ""),
    ("spider", "fly"),
    ("bird", "spider that wriggled and jiggled and tickled inside her"),
    ("cat", "bird"),
    ("dog", "cat"),
    ("goat", "dog"),
    ("cow", "goat"),
]

jokes = [
    "",
    "It wriggled and jiggled and tickled inside her.",
    "How absurd to swallow a bird!",
    "Imagine that, to swallow a cat!",
    "What a hog, to swallow a dog!",
    "Just opened her throat and swallowed a goat!",
    "I don't know how she swallowed a cow!",
]


def verse(num):
    if num == 7:
        return ["I know an old lady who swallowed a horse.", "She's dead, of course!"]

    res = [f"I know an old lady who swallowed a {animals[num][0]}."]
    if num > 0:
        res.append(jokes[num])

    for v in range(num, 0, -1):
        swallowed, to_catch = animals[v]
        res.append(f"She swallowed the {swallowed} to catch the {to_catch}.")

    res.append("I don't know why she swallowed the fly. Perhaps she'll die.")

    return res


def recite(start_verse, end_verse):
    verses = []

    for v in range(start_verse - 1, end_verse):
        verses.extend(verse(v))
        if v != end_verse - 1:
            verses.append("")

    return verses

File: solutions/go/food-chain/1/food_chain.go
package foodchain

import (
	"strings"
)

type verseDataObj struct {
	animalName string
	quip       string
	eatenName  string
}

var verseData = []verseDataObj{
	{animalName: "fly", quip: "I don't know why she swallowed the fly. Perhaps she'll die.", eatenName: ""},
	{animalName: "spider", quip: "It wriggled and jiggled and tickled inside her.", eatenName: "fly"},
	{animalName: "bird", quip: "How absurd to swallow a bird!", eatenName: "spider that wriggled and jiggled and tickled inside her"},
	{animalName: "cat", quip: "Imagine that, to swallow a cat!", eatenName: "bird"},
	{animalName: "dog", quip: "What a hog, to swallow a dog!", eatenName: "cat"},
	{animalName: "goat", quip: "Just opened her throat and swallowed a goat!", eatenName: "dog"},
	{animalName: "cow", quip: "I don't know how she swallowed a cow!", eatenName: "goat"},
	{animalName: "horse", quip: "She's dead, of course!", eatenName: ""},
}

// Verse returns the vth verse of the song.
func Verse(v int) string {
	var sb strings.Builder
	v -= 1

	sb.WriteString("I know an old lady who swallowed a " + verseData[v].animalName + ".\n" + verseData[v].quip)
	if verseData[v].eatenName != "" {
		for ; v > 0; v-- {
			sb.WriteString("\nShe swallowed the " + verseData[v].animalName + " to catch the " + verseData[v].eatenName + ".")
		}
		sb.WriteString("\n" + verseData[v].quip)
	}

	return sb.String()
}

// Verses returns the verses in the start to end range.
func Verses(start, end int) string {
	var sb strings.Builder

	for ; start <= end; start++ {
		sb.WriteString(Verse(start) + "\n\n")
	}

	return sb.String()[:sb.Len()-2]
}

// Song returns the whole song.
func Song() string {
	return Verses(1, 8)
}
